libexec/mass-testing/test-harness

```diff
@@ -0,0 +1,235 @@
+    with suppress(ProcessLookupError, PermissionError):
+        os.kill(pid, signal)
+
+
+def walk_process_tree(pid: int, include_root=True) -> Generator[psutil.Process, None, None]:
+    root_process = psutil.Process(pid)
+    if include_root:
+        pool = [root_process]
+    else:
+        pool = [*root_process.children()]
+
+    while len(pool) > 0:
+        new_pool = []
+        for element in pool:
+            yield element
+            new_pool.extend(element.children())
+        pool = new_pool
+
+
+def kill_and_reap_children(exclude_pids: Collection[int]):
+    # Kill all child processes and reap them. Returns True if any process was
+    # forcefully killed.
+    #
+    # NOTE: this creates a race, as new processes might be spawned in the
+    # meantime. An alternative might be to stuff the processes in a cgroupv2
+    # and use the 'cgroup.kill' special file
+    process_killed = False
+    while True:
+        children = [
+            p
+            for p in reversed(list(walk_process_tree(os.getpid(), False)))
+            if p.pid not in exclude_pids
+        ]
+        if len(children) == 0:
+            break
+
+        for process in children:
+            status = None
+            ppid = None
+            with suppress(psutil.NoSuchProcess):
+                status = process.status()
+            with suppress(psutil.NoSuchProcess):
+                ppid = process.ppid()
+
+            if status is None or ppid is None:
+                continue
+
+            if status not in (psutil.STATUS_ZOMBIE, psutil.STATUS_DEAD):
+                kill(process.pid, signal.SIGABRT)
+                process_killed = True
+                # Wait a bit between kills to allow for things to settle
+                sleep(2)
+            elif ppid == os.getpid():
+                os.waitpid(process.pid, 0)
+    return process_killed
+
+
+def pid_has_signal(pid: int, signalno: int) -> bool:
+    signal_bits = 0
+    with open(f"/proc/{pid}/status") as file:
+        for line in file:
+            if line.startswith("SigCgt:"):
+                signal_bits = int(line.removeprefix("SigCgt:"), 16)
+    return signal_bits & (2 ** (signalno - 1)) > 0
+
+
+def handle_timeout(root_process):
+    main_pid = None
+    for process in walk_process_tree(root_process.pid):
+        if pid_has_signal(process.pid, signal.SIGABRT):
+            main_pid = process.pid
+            break
+
+    if main_pid is None:
+        # Bail
+        root_process.send_signal(signal.SIGINT)
+        return
+
+    kill(main_pid, signal.SIGABRT)
+    try:
+        process.wait(30)
+    except psutil.TimeoutExpired:
+        kill(main_pid, signal.SIGKILL)
+        process.wait()
+    except psutil.ChildProcessError:
+        # Process died before we could `wait`
+        pass
+
+
+def patch_cmd(original_cmd: List[str], input_: str | None) -> List[str]:
+    result = []
+    for part in original_cmd:
+        if part == input_:
+            result.append("$INPUT")
```

Comment:
This is brittle and upon failure fails silently because `input_` is optional.
We can:
1. Make `--input` mandatory (making it positional) and have users do
   ```
   test-harness my-input-file -- my-command --lol %INPUT%
   ```
   Alternatively, we can drop `%INPUT%` and assume that the command you give accepts the input file as the last parameter.
2. Push the symlinking logic in `revng-test`, but I think that's not viable since the report generation code (which can only assume to know about this script basically) assumes the symlink is there.

Let me know if you have better ideas.

Reply:
1. I'd prefer that, it solves a whole bunch of problems. I'd use `%INPUT%` for added flexibility
2. Eh yes, I'd keep it in `test-harness`

Reply:
Let's go with #1.

Reply:
Implemented